### PR TITLE
🚧More notebook plotting backends: X3D and K3D

### LINKED ISSUE #258: K3D for Jupyter notebook rendering
We should look into using this toolset for some of the notebook rendering. From a  glance, it looks more robust than panel (what we currently use)

Take a look at the graphics in their README: https://github.com/K3D-tools/K3D-jupyter

### PR DESCRIPTION
**Pull request by banesullivan**
These changes allow users to choose a notebook plotting backend. This implements the use of X3D, VTKjs from `panel`, and K3D.

Everything here is a work in progress - especially the K3D implementation (it's rather hacky)

Ping @akaszynski - this is likely very helpful for the shading issues we're encountering with VTKjs

## Addresses

- Resolve https://github.com/pyvista/pyvista-support/issues/12 - scene will always output on `.show` call now
- Resolve #258 

### General

- [x] Make sure use of the old `use_panel` is switched to `nb_backend`
- [ ] Make sure docs and other repos like https://github.com/pyvista/cookiecutter-pyvista-binder are up to date with new `nb_viewer` option.
- [ ] Implement magic commands for #184 

### X3D

- [ ] X3D: Fix zoom scroll inversion issue
- [ ] X3D: Disable keyboard shortcuts in X3D scene
- [ ] X3D: Is there a way to embed a binary version of the HTML?
- [ ] X3D: Add more to the HTML to have buttons for viewing different camera angles, etc.

### K3D

- [ ] K3D: Fix issue with line sets (see https://github.com/K3D-tools/K3D-jupyter/issues/173)
- [ ] K3D: Set the color of text labels on scalar bars and axes from PyVsita's window/`rcParams`
- [ ] K3D: Add volume support with `vtkImageData`/`pyvista.UniformGrid`
- [ ] K3D: Fix point size and line width issue
- [ ] K3D: show edges of mesh?
- [ ] K3D: more informative names
- [ ] K3D: are textures possible?
- [ ] K3D: NaN colors?
- [ ] K3D: properly transfer lighting settings
- [ ] K3D: set camera position
- [ ] K3D: handle using RGBA arrays
- [ ] K3D: handle labeled points
- [ ] K3D: fix `n_colors` issue

## Differences

Check out the following example to see differences between the notebook backends and pure VTK rendering:

```py
import pyvista as pv
from pyvista import examples
m = examples.download_st_helens().warp_by_scalar()

cpos = [(567000.9232163235, 5119147.423216323, 6460.423216322832),
 (562835.0, 5114981.5, 2294.5),
 (-0.4082482904638299, -0.40824829046381844, 0.8164965809277649)]
```

```py
# Plot with X3D
pl = pv.Plotter(notebook=True)
pl.add_mesh(m, smooth_shading=True)
pl.show(nb_backend='x3d', cpos = cpos)
```

Screen Shot 2019-07-08 at 12 47 43 AM

```py
# Plot with VTKjs/panel
pl = pv.Plotter(notebook=True)
pl.add_mesh(m, smooth_shading=True)
pl.show(nb_backend='vtkjs', cpos=cpos)
```

Screen Shot 2019-07-08 at 12 47 49 AM

```py
# Plot with K3D
pl = pv.Plotter(notebook=True)
pl.add_mesh(m, smooth_shading=True)
pl.show(nb_backend='k3d')
```

Screen Shot 2019-07-08 at 6 19 03 PM

```py
# Plot with pure VTK
pl = pv.Plotter()
pl.add_mesh(m, smooth_shading=True)
pl.camera_position = cpos
pl.show(nb_backend=False)
```

download

### CODE REVIEW

**Comment by banesullivan** [comment]
Issue: scrolling is inverted - there's probably a way to specify this in the X3D HTML

**Comment by banesullivan** [comment]
Keyboard shortcuts for X3D:

- Press r to rest camera to starting position
- Press t to reset viewpoint.
- Press i to reset camera distance.
- Right-click and drag mouse to zoom.
- Middle-click and drag mouse to pan.

**Comment by banesullivan** [comment]
Okay so now I'm experiencing an issue where X3D only works with surface meshes if `smooth_shading` is enabled, otherwise, nothing is shown.

```py
import pyvista as pv
from pyvista import examples
m = examples.download_st_helens().warp_by_scalar()

cpos = [(567000.9232163235, 5119147.423216323, 6460.423216322832),
 (562835.0, 5114981.5, 2294.5),
 (-0.4082482904638299, -0.40824829046381844, 0.8164965809277649)]

# Plot with X3D
pl = pv.Plotter(notebook=True)
pl.add_mesh(m.contour(), color='black')
pl.add_mesh(m, smooth_shading=False)
pl.show(nb_backend='x3d', cpos = cpos)
```

Screen Shot 2019-07-08 at 10 26 45 AM

**Comment by banesullivan** [comment]
Also solid colors do not seem to work. In this example i try to plot white but it's black:

```py
import pyvista as pv
from pyvista import examples
m = examples.download_st_helens().warp_by_scalar()

cpos = [(567000.9232163235, 5119147.423216323, 6460.423216322832),
 (562835.0, 5114981.5, 2294.5),
 (-0.4082482904638299, -0.40824829046381844, 0.8164965809277649)]

# Plot with X3D
pl = pv.Plotter(notebook=True)
pl.add_mesh(m.contour(), color='white')
pl.show(nb_backend='x3d', cpos = cpos)
```

Screen Shot 2019-07-08 at 10 27 36 AM

**Comment by banesullivan** [comment]
Appearantly VTK's X3D exporter `vtk.vtkX3DExporter` which we are using has a lot of issues... Maybe we should/could make our own exporter that follows the X3D specification. Considering how widely used X3D is, this might not be a huge task.

Also... X3D allows the use of scalar bars, widgets, and more. See [this demo](http://fpavogt.github.io/x3d-pathway/dices.html) for more details

**Comment by banesullivan** [comment]
From the comparison in the main description here, K3D seems to do the best at mimicking VTK's lighting parameters

**Comment by GuillaumeFavelier** [comment]
This is amazing work! I can't wait to try it out myself. I particularly appreciate how X3D tackle transparency issues with **transparency layers**, I wonder how K3D does it, I will try to find out.

The [integration with VTK](https://k3d-jupyter.readthedocs.io/en/latest/showcase/VTK_cutter.html) in K3D is impressive, the controls by widgets is very useful too.

**Comment by flutefreak7** [comment]
Just wanted to say this looks amazing! Keep up the good work!

**Comment by banesullivan** [comment]
Thanks for the feedback, @flutefreak7! 

Unfortunately, this PR is a bit stale and will likely remain delayed while I focus on implementing a better solution using [`itkwidgets`](https://github.com/InsightSoftwareConsortium/itk-jupyter-widgets) - reference https://github.com/InsightSoftwareConsortium/itk-jupyter-widgets/pull/154 and linked issues/PRs

File: pyvista/plotting/plotting.py [hunk 1]
```diff
@@ -3757,28 +3786,39 @@ def show(self, title=None, window_size=None, interactive=True,
...
-            use_panel = rcParams['use_panel']
-
+        if self.notebook:
+            try:
+                import IPython
+                from IPython.display import HTML
+            except ImportError:
+                raise Exception('Install IPython to display image in a notebook')
+        if use_panel:
+            warnings.warning("`use_panel` is deprecated. Use `nb_backend` insted!!")
```
Comment (banesullivan) [inline pyvista/plotting/plotting.py]:
```suggestion
            warnings.warning("`use_panel` is deprecated. Use `nb_backend` instead!!")
```

**Comment by akaszynski** [comment]
@banesullivan,

This is an old PR, but I think given how some people (including myself) are having problems with `itkwidgets`, it would be great to have multiple options for plotting within jupyter.  Plotting within a notebook, especially on a remote server, is becoming more popular these days and `pyvista` needs to have an interface for these backends.  K3D seems as popular as `itkwidgets`, and `panel` is hugely popular as well.

We should consider creating multiple PRs for each of these tools rather than one huge one.  In my (limited) experience, the bigger the PR the longer it takes for it to get merged.

**Comment by banesullivan** [comment]
@akaszynski, agreed. Let's keep this open until I have the time to come back to this... shouldn't be terribly long

**Comment by agramfort** [comment]
> I think given how some people (including myself) are having problems with
itkwidgets,

did you try to report the issues upstream with itkwidgets devs? Offering
too many options
is also dangerous as it scatters our dev ressources. my 2c

>

**Comment by akaszynski** [comment]
I ended up being able fix the problem with a full reinstall, so the issue was likely on my end.

You’re correct about our limited resources being scattered, and I feel like we’re currently already quite busy with existing issues, not to mention our day jobs :)

As different back ends have different advantages, it’s tough to say at the moment which one to support.  Itkwidgets doesn’t render on the iPhone or iPad, making it a tough sell for one of my closed source projects, while `panel` does.

For the time being, I think minimal support for a few backends should be a short term goal, and then we might consider supporting the lead backend should users desire it.

**Comment by banesullivan** [comment]
This PR has gone quite stale and I am no longer going to pursue K3D or X3D as viable notebook plotting backends. For a pure web-based rendering solution, a vtkjs-based approach is a clear winner in my mind. This approach could come from `itkwdigets`, us making our own implementation in JS via a synchronization between the VTK render window and a vtkjs one (see #698), or from the awesome work going on lately in `panel` (see https://github.com/pyvista/pyvista/pull/268#issuecomment-619575438).

---

Maybe someone will find the X3D export script useful and that could be implemented in a new PR down the road if someone needs it.

As for K3D, far too much work is required to have a working solution.

**Pull request closed without merging**